### ISSUE TITLE
Interruption in jeromq kills it

### ISSUE THREAD

**Issue by fbacchella** [zeromq/jeromq#834]
It's known that jeromq don't like interruptions. But it hate it so much that an interruption at wrong time can kill it:

```
    @Test(timeout = 1000)
    public void testInterruptedClose() throws Throwable
    {
        ZContext ctx = new ZContext();
        Socket s = ctx.createSocket(SocketType.PULL);
        int port = Utils.findOpenPort();
        s.bind("tcp://*:" + port);
        Thread.currentThread().interrupt();
        try {
            s.close();
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        // Reset the interrupt state
        Thread.interrupted();
        ctx.close();
    }
```

An exception is of course throws during the close. But as the close is not really finished, the ctx.close will hang, the test fails with a timeout during the zmq.Ctx.terminate().

Is such a catastrophic failure expected ?

**Comment by trevorbernard** [comment]
Ideally it shouldn't fail in a ball of flames but causing an interrupt is an undetermined behaviour. We really don't know/understand the consequences of interrupting that thread.

**Comment by fbacchella** [comment]
I'm OK with the current thread failings. But I spent hours trying to understand why latter, in another part of my code, context termination was hanging.
Il will need to add a lot of guard along around my code to prevent jeromq to interrupt.

**Comment by trevorbernard** [comment]
I've attempted to totally wrap my head around it before but unfortunately, it's a bit too much effort and low on my personal priority list. The cognitively complexity of the internals of libzmq/jeromq is quite large

**Comment by fbacchella** [comment]
I tried a simple but not smart solution. Interruptions are masked and it retries:
```
diff --git a/src/main/java/zmq/Signaler.java b/src/main/java/zmq/Signaler.java
index e9af374..b66a4a1 100644
--- a/src/main/java/zmq/Signaler.java
+++ b/src/main/java/zmq/Signaler.java
@@ -3,6 +3,7 @@ package zmq;
 import java.io.Closeable;
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.nio.channels.ClosedByInterruptException;
 import java.nio.channels.ClosedChannelException;
 import java.nio.channels.ClosedSelectorException;
 import java.nio.channels.Pipe;
@@ -90,12 +91,27 @@ final class Signaler implements Closeable
 
     void send()
     {
-        int nbytes;
+        int nbytes = 0;
 
         while (true) {
             try {
                 wdummy.clear();
-                nbytes = w.write(wdummy);
+                boolean wrote = false;
+                // This loop try to protect the current thread from external interruption.
+                // If it happens, it brake current context internal state, but try
+                // anyway to keep memory of the event.
+                boolean interrupted = Thread.interrupted();
+                while (! wrote) {
+                    try {
+                        nbytes = w.write(wdummy);
+                        wrote = true;
+                    } catch (ClosedByInterruptException e) {
+                        interrupted = true;
+                    }
+                }
+                if (interrupted) {
+                    Thread.currentThread().interrupt();
+                }
             }
             catch (IOException e) {
                 throw new ZError.IOException(e);
```

**Comment by fbacchella** [comment]
Fixed in #836
